feat: associates user by email for oauth when tpa is required

nizarmah opened the pull request:
## Description

Currently, in the edX platform, if you try to login with an SSO account using the login form, a username is generated for you based on the email that is used with the SSO provider.

The username is generated using the pipeline `third_party_auth.pipeline.get_username`.

If the username already exists in the platform, you run into a small conflict which sends you back to the login form and requests the following:
![image](https://user-images.githubusercontent.com/5631091/109410595-f82d8280-79ac-11eb-854f-d60339e9fd9b.png)

It basically requests that you link your edX account and your SSO account.

However, sometimes we would want the edX account and SSO account to be associated by email. There's already a special pipeline available in the platform for that, `third_party_auth.pipeline.associate_by_email_if_login_api`. Sadly, however, it is only available for a certain entry method of authentication, called `login_api`.

~~So in order to start associating the edX account and SSO account by email, this pull request adds a waffle switch.~~

~~**Update:** After a lot of reading through the code and investigation, it seems like the `login_api` auth entry is no longer used. Accordingly, I took the liberty of removing said pipeline and replacing it with something that seemed useful. Now the pipeline runs whenever third party authentication is required.~~

**Update**: After @waheedahmed's [findings](https://github.com/edx/edx-platform/pull/25935#pullrequestreview-611983769), which is that `login_api` is in fact still used, and @zamanafzal's [request](https://github.com/edx/edx-platform/pull/25935#discussion_r595030485), which is to make the functionality specific to Oauth providers, a new pipeline was added to associate the user by email **only when the following conditions are met**:
* The social auth provider is an Oauth2 provider
* The `ENABLE_REQUIRE_THIRD_PARTY_AUTH` is enabled

## Supporting information

* **JIRA tickets**: SE-3816, [OSPR-5312](https://openedx.atlassian.net/browse/OSPR-5312)
* **Sandbox URL**:
  * Default Behavior:
    * LMS: https://pr25935-default.sandbox.opencraft.hosting/
    * Studio: https://studio.pr25935-default.sandbox.opencraft.hosting/
  * New Behavior:
    * LMS: https://pr25935.sandbox.opencraft.hosting/
    * Studio: https://studio.pr25935.sandbox.opencraft.hosting/

## Testing instructions

### Preparation

Please send an email to `nizar` at `opencraft.com`, requesting to be added as a test user to the Google SSO used for this sandbox.

**You need a Google Account, because you'll be using the same email address when registering through the form and through the SSO.**

### Reproducing the problem

#### Logging in with the SSO

Login with the Google SSO over at the sandbox with [default behavior](https://pr25935-default.sandbox.opencraft.hosting/)

#### Removing Social Auth Record

Sign in as `staff`:`edx` into the [django admin](https://pr25935-default.sandbox.opencraft.hosting/admin) and delete the [user social auth](https://pr25935-default.sandbox.opencraft.hosting/admin/social_django/usersocialauth) for your user.

#### Logging in with the SSO, again...

Login, again, with the Google SSO over at the sandbox with [default behavior](https://pr25935-default.sandbox.opencraft.hosting/). After doing that, you should receive a similar screenshot
![image](https://user-images.githubusercontent.com/5631091/109410595-f82d8280-79ac-11eb-854f-d60339e9fd9b.png)

### Testing the fix

#### Logging in with the SSO

Login with the Google SSO over at the sandbox with [default behavior](https://pr25935.sandbox.opencraft.hosting/)

#### Removing Social Auth Record

Sign in as `staff`:`edx` into the [django admin](https://pr25935.sandbox.opencraft.hosting/admin) and delete the [user social auth](https://pr25935.sandbox.opencraft.hosting/admin/social_django/usersocialauth) for your user.

#### Logging in with the SSO, again...

Login, again, with the Google SSO over at the sandbox with [default behavior](https://pr25935.sandbox.opencraft.hosting/). After that, you should be able to directly login 😃 

**Settings**
```yaml
EDXAPP_LMS_ENV_EXTRA:
  ENABLE_REQUIRE_THIRD_PARTY_AUTH: true
```

edx-webhook commented:
Thanks for the pull request, @nizarmah! I've created [OSPR-5312](https://openedx.atlassian.net/browse/OSPR-5312) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

nizarmah commented:
Sorry for all the messy force pushes! I just wanted to clean the history before any reviewer took a look. I have been testing on the sandbox, which made pushing to the branch just more convenient.

gabor-boros commented:
👍  🎉 

- [x] I tested what's in the test instructions
- [x] I read through the code
- NA I checked for accessibility issues
- NA Includes documentation
- NA I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

nizarmah commented:
@natabene, I hope you and the members on the edX team are having/have had a nice holiday! 
This is ready for edX's review 👍🏼

natabene commented:
@nizarmah Thanks, lining this up for our review.

nizarmah commented:
Please note that this failed build isn't caused by my changes. It seems like an issue that's common among different branches, including master.

bradenmacdonald commented:
This will likely need product review but I would be inclined to suggest making this the default behavior and deprecating the previous behavior with the yellow message ("You have successfully signed in but there's no linked account"). Everyone I've ever talked to about the current flow has said it's confusing.

The most common case is that the user's SSO email matches their existing account's email, and in that case it should automatically link (if we trust the SSO provider), which I believe is what's happening here.

The "edge case" is that the user's SSO email is different from their existing account's email; in that case if they want to "link" instead of creating a new account, they can always sign in to their existing account, go to the account settings page, and link the accounts from there. That's how most other websites work anyways.

natabene commented:
@nasthagiri Are you ok with this going to core committers?

nasthagiri commented:
@natabene Yes, I am.

@bradenmacdonald +1 to your suggestion of changing the default.  @wajeeha-khalid can provide Product input on this.

bradenmacdonald commented:
Oh, hi @wajeeha-khalid 😊 Let me know your thoughts on this, or if you want to chat about it.

wajeeha-khalid commented:
@bradenmacdonald hey there 😊
I can agree no more, the existing SSO behavior where user lands on the login page with a warning message is confusing.
Work in this PR can be of great value and your suggestion of making it default seems like a good idea.

Give me some time to review what's available after this change in the sandbox linked above and also loop in my team to get their input on this.

As an FYI, we are currently migrating Logistration (login, register, reset password) onto its own MFE and we are hoping to launch that hopefully sometime in February. Another initiative, right after the MFE launch is the re-design of these pages and our UX team is currently working on finalizing those new designs.

I would like us to think about whether we would want to combine this change with any of the other two changes or we can also take it separately pre or post MFE launch.

nizarmah commented:
> Give me some time to review what's available after this change in the sandbox linked above and also loop in my team to get their input on this.

@wajeeha-khalid please let me know if you'd like me to add specific email addresses as testers for the Google SSO that is used in the sandbox 😃

wajeeha-khalid commented:
@nizarmah yes, please add these three emails; it will help me test the use-case.

- wajeeha.khalid@arbisoft.com
- jia.khalid7@gmail.com
- waheed.ahmed@arbisoft.com

nizarmah commented:
@wajeeha-khalid the email addresses you mentioned have been added! 
Please let me know if there's anything else that I can assist with 🙂

wajeeha-khalid commented:
@nizarmah I have reviewed the behavior and it seems great 👍 Thank you for making this change.
I have asked @waheedahmed (who is the technical lead for edX's Logistration features) to take a look at it; so he is also aware of this change.

I have two questions.
1.  @waheedahmed once we launch MFE (in the next couple of weeks), if this PR gets merged, will this change become available on MFE too or we would need to do some work on MFE to make it available?
2. @nizarmah if we decide to make it a default behavior; would we need the waffle switch then?

waheedahmed reviewed (approved):
@wajeeha-khalid I have tested it thoroughly with and without logistration MFE and I believe we don't need any further changes for our MFE or monolith FE except just enabling the waffle flag if we don’t choose to enable it by default.

The changes look good to me 👍, thank you!

nizarmah commented:
@wajeeha-khalid 

> if we decide to make it a default behavior; would we need the waffle switch then?

If it is a default behavior, then the waffle switch would not be needed. But to that I have some follow-up questions, which are mainly related to @bradenmacdonald's suggestions.

> This will likely need product review but I would be inclined to suggest making this the default behavior and deprecating the previous behavior with the yellow message ("You have successfully signed in but there's no linked account"). Everyone I've ever talked to about the current flow has said it's confusing.

Should I update the pull request to remove all the code related to this deprecated behavior, where the yellow message shows up?

> It may also be necessary to enable this per provider, because we don't necessarily trust every provider's "email" field (e.g. user logs in via their company provider and company provider was hacked to say that they own the email of an edX employee, we don't want to log them in to the edX employee's account).

Would you prefer I create this as a default behavior, per provider, and add an option to override it and disable it, _also_ per provider? This would be done mainly through the _Other Settings_ of the added third party provider (oauth).

---

I won't work on any changes unless you sign off on them first, so if you have something else in mind, please let me know 😃

wajeeha-khalid commented:
@nizarmah I think for now we can merge it as is (not making it default). I can take this to our UX team as well and plan to include this (making it default) as part of our re-design initiative; unless @bradenmacdonald you have a different opinion.

bradenmacdonald commented:
@wajeeha-khalid That makes sense to me! There's no point in spending more time re-working this if you're going to be redesigning it all anyways.

natabene commented:
@waheedahmed @bradenmacdonald Thank you both. So will this PR be merged or closed?

nizarmah commented on common/djangoapps/third_party_auth/config/waffle.py:
```diff
@@ -0,0 +1,24 @@
+
+WAFFLE_NAMESPACE = 'third_party_auth'
+
+# .. toggle_name: ALWAYS_ASSOCIATE_USER_BY_EMAIL
```
```suggestion
# .. toggle_name: third_party_auth.always_associate_user_by_email
```

nizarmah commented on common/djangoapps/third_party_auth/config/waffle.py:
```diff
@@ -0,0 +1,24 @@
+"""
+
+
+from edx_toggles.toggles.__future__ import WaffleSwitch
```
```suggestion
from edx_toggles.toggles import WaffleSwitch
```

nizarmah commented:
Of course... I should have rebased 🤦🏼

natabene commented:
@wajeeha-khalid Ready to merge you think?

wajeeha-khalid commented:
yes 👍

waheedahmed commented:
@natabene Yes, thank you!

nizarmah commented:
That's awesome to hear! Sorry for the conflicts, I'll squash and resolve them by the end of the day.

Thanks for your review @wajeeha-khalid @waheedahmed. Hope you have a wonderful week 😃

zamanafzal commented:
Please wait on merging this. A similar kind of work has been done in this [PR](https://github.com/edx/edx-platform/pull/26170) which has been merged earlier today.

nizarmah commented:
@zamanafzal the changes in this PR are intended in order to associate the email for other providers. Would you like me to combine those two methods into a single one instead?

zamanafzal commented:
@nizarmah Your changes will affect the users coming from the SAML provider. As after your check, we are enabling the default behavior of associate_by_email which is not desired in the SAML SSO authentication case. Also, as per my knowledge, this [check](https://github.com/edx/edx-platform/blob/e9dc5baf79e50f42a64a69832f67a6945b07819d/common/djangoapps/third_party_auth/pipeline.py#L727)
allows all oauth2 provider users to associate their email with existing accounts and google is also an oauth2 provider. 
FYI: @moconnell1453

robrap commented on common/djangoapps/third_party_auth/config/waffle.py:
```diff
@@ -18,8 +17,21 @@
 # .. toggle_target_removal_date: 2021-04-31
 # .. toggle_warnings: None.
```
I haven't gotten to review yet, but do you mind deleting `toggle_warnings` from this older annotation? Also, I think the second line of the `toggle_description` here will need more indents to not fail later.  Thank you!

robrap commented on common/djangoapps/third_party_auth/config/waffle.py:
```diff
@@ -18,8 +17,21 @@
+# .. toggle_description: Always associates current social auth user with
+#   the user with the same email address in the database, which verifies
+#   that only a single database user is associated with the email.
+# .. toggle_use_cases: opt_in
```
@nizarmah: When I wrote [my last comment](https://github.com/edx/edx-platform/pull/25935/files#r567959945), I thought this was part of the docathon effort.  The comment still stands, but I didn't realize this was already fully reviewed, so I don't plan on reviewing.

However, according the the review comments, it seems like this toggle is `temporary`, and just being used for rollout to ultimately switch over the the newer way.  I may have misunderstood though.  If this is true, please switch to `temporary` and maybe note that this is just being used for rollout. If I misunderstood the other comments on the PR, you can correct me and ignore this. Thanks.

zamanafzal commented on common/djangoapps/third_party_auth/config/waffle.py:
```diff
@@ -18,8 +17,21 @@
-    __name__
+ENABLE_MULTIPLE_SSO_ACCOUNTS_ASSOCIATION_TO_SAML_USER = WaffleSwitch(
+    f'{WAFFLE_NAMESPACE}.enable_multiple_sso_accounts_association_to_saml_user',
+    module_name=__name__
```
@nizarmah Thank you for taking care of this. I am creating another PR for this waffle switch change. Also one NIT here, we are declaring the variable in the string but not passing value to it. I think it should be
`'{WAFFLE_NAMESPACE}.enable_multiple_sso_accounts_association_to_saml_user'.format(WAFFLE_NAMESPACE=WAFFLE_NAMESPACE)`

nizarmah replied:
@zamanafzal the f-strings in python is really helpful in that regard. We don't need to add the `.format` part 😃

nizarmah commented:
@zamanafzal the condition you shared is related to the `login_api`, which doesn't seem to be used at the moment with oauth.
I added a sandbox above, which you can disable and enable the flag that I added in order to verify that previously the SSO account wasn't being associated with the email. Therefore, this flag is needed, unless I'm doing something wrong on my end, please let me know.

cc @wajeeha-khalid @waheedahmed (ccing reviewers on the pull request because they might have a better justification as to why this flag is needed)

natabene commented:
@ziafazal @wajeeha-khalid @bradenmacdonald What is the current status of this PR? Is merging still blocked by anything?

nizarmah commented:
@natabene I think it is blocked at me better describing the need of this pull request, and how the comment @zamanafzal left isn't the case for Oauth providers (at least based on my interactions with the oauth provider). I plan to address that soon, but I've been a little busy with something else with higher priority at the moment. Sorry about that.

natabene commented:
@nizarmah No worries, thank you for clarifying for me, I will change the status accordingly.

nizarmah commented:
Small update here: I'm still looking for a different solution. It seems like [the line shared](https://github.com/edx/edx-platform/pull/25935#issuecomment-770943720) by @zamanafzal, `auth_entry == AUTH_ENTRY_LOGIN_API`, is not being used anymore. 

I also looked into `/login_ajax` which is also known as `login_api`. Unfortunately, it also seems like this isn't used anymore.

This will take some additional time.

@zamanafzal if you could provide a more detailed up-to-date reply regarding your [reply earlier](https://github.com/edx/edx-platform/pull/25935#issuecomment-770943720), it would help find a middle grounds for that change quicker.

cc @waheedahmed if you have any comments that might help, please let me know.

nizarmah commented:
By the way, I changed the `login_api` auth entry condition for the email association pipeline. The `login_api` auth entry is no longer used.

In addition, when `` is enabled, no login form shows up. So the following error message doesn't make sense:
![image](https://user-images.githubusercontent.com/5631091/109410637-4478c280-79ad-11eb-8b3c-55165da42145.png)

The accounts (edX and SSO accounts) should directly be associated by the user's emails, because the user can't even link the two accounts together because the form doesn't render anyway.

wajeeha-khalid commented:
@nizarmah An FYI, we are launching MFE for edX today and we would like to monitor our key metrics for this launch for at least 1 week. I would like no other intervention that can affect our registration rate and login rate during this time.
I would appreciate it if you can hold on to merging this change for at least one week.
cc: @waheedahmed

nizarmah commented:
@wajeeha-khalid thanks for the update 😃 Hope the MFE launch went as planned!

I was wondering, is the existing approval on the pull request still enough, after the changes that were done, or is another engineering review needed? If so, would @waheedahmed be able to check the changes?

natabene commented:
@waheedahmed Could you please check and merge if all looks good?

bradenmacdonald commented:
@natabene Sure, I can do that if @waheedahmed isn't available now.

waheedahmed commented:
@natabene I'll take a look at it today and will merge it if all looks good.

CC: @wajeeha-khalid @bradenmacdonald

waheedahmed commented:
@nizarmah I suspect this auth entry is still in use, I'm trying to verify it and I'm afraid it won't get merged until I have concrete evidence.

nizarmah commented:
@waheedahmed I totally understand. Take your time. 

I spent a lot of hours trying to verify that it is still in use, but I had no luck. I'm sure you'll have a better chance at finding concrete evidence though, good luck.

I'll try to see if I can reserve some time next week to try and look further into it and see if I can find anything concrete.

waheedahmed commented:
@nizarmah I'm adding a custom attribute with auth entry value on New Relic transactions. Hopefully, it will help us verifying it without any code exploration.

waheedahmed reviewed (changes requested):
Hello @nizarmah, seems like this auth entry `login_api` is still widely used within edX, and most probably by the open community as well, following are the results for just half an hour:
<img width="1092" alt="Screenshot 2021-03-15 at 2 26 28 PM" src="https://user-images.githubusercontent.com/2851134/111132220-e518d600-859a-11eb-96cd-78d87dc0f37c.png">

Could you please check and make the changes accordingly?

zamanafzal commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -724,25 +727,20 @@ def login_analytics(strategy, auth_entry, current_partial=None, *args, **kwargs)
 
 @partial.partial
-def associate_by_email_if_login_api(auth_entry, backend, details, user, current_partial=None, *args, **kwargs):  # lint-amnesty, pylint: disable=keyword-arg-before-vararg
+def associate_by_email_if_third_party_auth_is_required(auth_entry, backend, details, user, strategy, *args, **kwargs):
```
Also, Please make this code specific to OAuth functionality. As currently it is generic and will impact the already existing SAML flow for this part. Please take a look at the following link.
cc: @waheedahmed @moconnell1453 
https://github.com/edx/edx-platform/blob/0ca2295e8f8a953790cca8f0fe09787f2be9c408/common/djangoapps/third_party_auth/pipeline.py#L756

nizarmah commented:
@waheedahmed @zamanafzal hello! sorry for the delayed reply! I will be sure to address both of your comments tomorrow.

By the way @waheedahmed, thanks for the screenshot. I'm really glad you shared that with me!

nizarmah commented:
jenkins run python

nizarmah commented:
@waheedahmed I'm adjusting the `is_oauth_provider` since it's still not recognizing properly the Oauth2 backends. I'll let you know once the sandbox is working as expected. 👍🏼

nizarmah commented:
@waheedahmed @zamanafzal this is ready for your review 😃  The sandbox is working as well. 

cc @natabene

waheedahmed commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -739,16 +740,29 @@ def associate_by_email_if_login_api(auth_entry, backend, details, user, current_
+    This association is done ONLY if the user entered the pipeline belongs to Oauth provider.
+    """
+
+    if is_oauth_provider(backend.name, **kwargs):
```
It will enable it by default for each auth entry in the oauth2 provider's case and I believe the purpose of `associate_by_email_if_login_api` is failed here. Don’t you think we need to put it under the feature flag `ENABLE_REQUIRE_THIRD_PARTY_AUTH`?

waheedahmed replied:
And if we are making it the default behavior in the oauth2 providers case, then I think we can change this check `if auth_entry == AUTH_ENTRY_LOGIN_API:` to something `if is_oauth_provider(backend.name, **kwargs) and auth_entry in ALLOWED_AUTH_ENTRIES:` instead of this new pipeline function.

`ALLOWED_AUTH_ENTRIES` will be a list of auth entries you want to allow here. e.g. `login_api`, `login`, `register`.

nizarmah replied:
@waheedahmed Yes, I do think we need to put it under the feature flag `ENABLE_REQUIRE_THIRD_PARTY_AUTH`, that's a great point.

Your recommendation about the `ALLOWED_AUTH_ENTRIES` is great, but I'm hesitant to make such a change because it can be considered "breaking".

nizarmah replied:
@waheedahmed may you please check the latest changes? I added the setting check. I also added a test to verify that it is working 😃 

I thought of adding another test, when the flag is disabled, but I believe `test_new_account_registration_assigns_distinct_username_on_collision` already covers the default case. Let me know though, please, if you don't consider that to be enough 👍🏼

waheedahmed reviewed (commented):
Just one concern otherwise looks good now.

waheedahmed commented on common/djangoapps/third_party_auth/tests/specs/base.py:
```diff
@@ -906,6 +906,30 @@ def test_new_account_registration_assigns_distinct_username_on_collision(self):
         distinct_username = pipeline.get(request)['kwargs']['username']
         assert original_username != distinct_username
 
+    @django_utils.override_settings(ENABLE_REQUIRE_THIRD_PARTY_AUTH=True)
```
This test is passing even with `ENABLE_REQUIRE_THIRD_PARTY_AUTH=False`, shouldn’t it fail without this setting enabled?

nizarmah replied:
😯 it should indeed fail without this setting enabled... I'll look into it. Sorry about that!

nizarmah commented:
@waheedahmed sorry for this taking so much longer than it should 😞 

Anyway, I updated the tests, and I also added some additional tests for the utils I added 👍🏼 
I made sure to test everything extensively this time, instead of rushing it. 

I went also with a different approach for the test, which is just the opposite of `test_signin_fails_if_no_account_associated`. 
I just had to mock certain stuff because the pipeline made things pretty difficult.

This should be ready for your review again 😁 

cc @natabene

robrap commented:
For historical purposes, would you mind updating the PR description? Thank you.

nizarmah commented:
@robrap thanks for the reminder 👍🏼 The description has been updated, and I made sure to link to two important messages in the pull request. 😄

waheedahmed reviewed (approved):
A small nit otherwise looks good now 👍, please squash your commits and let me know when it's ready to merge.

Thanks!

nizarmah commented on common/djangoapps/third_party_auth/tests/specs/base.py:
```diff
@@ -789,6 +789,27 @@ def test_signin_fails_if_no_account_associated(self):
+        with mock.patch(
+            'common.djangoapps.third_party_auth.pipeline.get_associated_user_by_email_response',
+            return_value=[{'user': user}, True],
+        ) as _mock_get_associated_user_by_email_response:
```
Sure! 👍🏼 

I also removed `_mock_associate_by_email` since it was also not used 😃 

This should be addressed by now. I also squashed by the way.

nizarmah commented:
@waheedahmed 🥳 awesome! 

Thank you a lot for the review and for being patient with me through this change 😅 

I addressed your nit and squashed. This is ready to be merged 👍🏼

edx-webhook commented:
@nizarmah 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/25935&entry.1484655318=https://github.com/nizarmah&entry.752974735=2020-12-23+05:23&entry.1917517419=2021-04-02+06:32&entry.2133058324=Yes) so we can improve your experience in the future.
